### PR TITLE
Course completed event creation

### PR DESCRIPTION
**Pull request by caperneoignis**
This resolves issue #29 (course completed half) and allows for course completed. Please let me know if any issues need to corrected. This passed all test locally in my phpunit. Feel free to give feed back if any issues need to be resolved.

### CODE REVIEW

**Comment by ryansmith94** [comment]
Looks ok to me @jlowe64.

File: src/Events/CourseCompleted.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+     * @override Event
+     */
+    public function read(array $opts) {
+        $version = str_replace("\n", "", str_replace("\r", "", file_get_contents(__DIR__.'/../../VERSION')));
+        return [
+            'user' => $opts['relateduserid'] < 1 ? null : $this->repo->readUser($opts['relateduserid']),
+            'relateduser' => $opts['relateduserid'] < 1 ? null : $this->repo->readUser($opts['relateduserid']),
+            'course' => $this->repo->readCourse($opts['courseid']),
+            'app' => $this->repo->readCourse(1),
+            'info' => (object) [
```
Comment (ryansmith94) [inline src/Events/CourseCompleted.php]: This stuff should be necessary. Look at the [ModuleEvent](https://github.com/LearningLocker/Moodle-Log-Expander/blob/master/src/Events/ModuleEvent.php).

**Comment by ryansmith94** [comment]
Actually @jlowe64, I do have a comment. Once that is resolved I think it should be ok.

File: src/Controller.php [hunk 2]
```diff
@@ -52,7 +51,22 @@ public function __construct(Repository $repo) {
     }
 
     /**
-     * Creates new events.
+     * Creates a new event.
```
Comment (ryansmith94) [inline src/Controller.php]: This needs to be removed. I think this was accidentally added.

File: src/Controller.php [hunk 3]
```diff
@@ -61,13 +75,8 @@ public function createEvents(array $events) {
         foreach ($events as $index => $opts) {
             $route = isset($opts['eventname']) ? $opts['eventname'] : '';
             if (isset(static::$routes[$route]) && ($opts['userid'] > 0 || $opts['relateduserid'] > 0)) {
-                try {
```
Comment (ryansmith94) [inline src/Controller.php]: This needs to be reverted too.

File: src/Events/CourseCompleted.php (hunk 1 above)
Comment (caperneoignis) [inline src/Events/CourseCompleted.php]: Should or should not? Since it extends event I think we can get away with out having most of that, and just do a return  parent with some of the conversions? I think that's how I had orignally.

Comment (caperneoignis) [inline src/Events/CourseCompleted.php]: I see what your getting at, I just clicked on the link. Okay I'll convert it back with parent::read, like I had it orignally.

Comment (ryansmith94) [inline src/Events/CourseCompleted.php]: Sorry "should not". Awesome thanks @caperneoignis 👍

**Comment by caperneoignis** [comment]
Okay, real quick, for these items where I need to make changes, do I just make another pull request? or is there a way to modify this request so it updates with the changes? I'm new to this, believe it or not, I'm the only programmer in my department, so I'm the only one that really uses our git server. lol Any way let me know, I'll try and make the changes to the event today after lunch and resubmit after running the test.

**Comment by jlowe64** [comment]
You just commit to your repo as normal and it'll populate over the changes here.

**Comment by caperneoignis** [comment]
I have fixed the issues, that @ryansmith94 had left. Let me know if you have any questions. I'll start working on the plugin test tomorrow.

File: src/Controller.php [hunk 4]
```diff
@@ -26,9 +26,8 @@ class Controller extends PhpObj {
         '\mod_glossary\event\course_module_viewed' => 'ModuleEvent',
         '\mod_imscp\event\course_module_viewed' => 'ModuleEvent',
         '\mod_survey\event\course_module_viewed' => 'ModuleEvent',
-        '\mod_url\event\course_module_viewed' => 'ModuleEvent',
+        //'\mod_url\event\course_module_viewed' => 'ModuleEvent',
```
Comment (ryansmith94) [inline src/Controller.php]: Why is this commented out?

File: src/Controller.php [hunk 5]
```diff
@@ -26,9 +26,8 @@ class Controller extends PhpObj {
...
         '\mod_facetoface\event\course_module_viewed' => 'ModuleEvent',
-        '\mod_feedback\event\course_module_viewed' => 'ModuleEvent',
```
Comment (ryansmith94) [inline src/Controller.php]: Why was this removed?

File: src/Controller.php [hunk 6]
```diff
@@ -40,7 +39,7 @@ class Controller extends PhpObj {
         '\core\event\user_created' => 'Event',
         '\core\event\user_enrolment_created' => 'Event',
         '\mod_scorm\event\sco_launched' => 'ScormLaunched',
-        '\mod_feedback\event\response_submitted' => 'FeedbackSubmitted'
```
Comment (ryansmith94) [inline src/Controller.php]: Why was this removed?

**Comment by ryansmith94** [comment]
Added new comments to the code @caperneoignis

File: src/Controller.php (hunk 4 above)
Comment (caperneoignis) [inline src/Controller.php]: That is commented out, because it's the same key twice. Wasn't sure why we have the same key twice on that one.

File: src/Controller.php (hunk 5 above)
Comment (caperneoignis) [inline src/Controller.php]: Great question. I don't remember removing it. I must have a phantom in my keyboard, because ive noticed that once or twice where something magically got deleted, when  I was not looking. I'll add that back, but I have no clue why it got removed.

File: src/Controller.php (hunk 6 above)
Comment (caperneoignis) [inline src/Controller.php]: Same as above, not sure why or how that got deleted. The only one I was tracking was the commented out one. I must have fat fingered it. I'll fix those issues. Thanks for spotting my mess ups. I'll look harder had the changes.

**Comment by caperneoignis** [comment]
Okay, I just copied the controller.php code, and added the course completed event back to the routing array. However; these two keys are duplicates
````javascript
        '\mod_url\event\course_module_viewed' => 'ModuleEvent',
        '\mod_feedback\event\course_module_viewed' => 'ModuleEvent',
````

File: src/Controller.php (hunk 6 above)
Comment (caperneoignis) [inline src/Controller.php]: Okay, I added the item back, but for some reason it is not showing this as an out of date diff. But I just checked it's in my code.

Comment (ryansmith94) [inline src/Controller.php]: It'll be because of the comma that's been added at the end of the line, which is why I usually add a comma to the end of the last element too in PHP code to avoid this situation in the diff.

Comment (caperneoignis) [inline src/Controller.php]: Awww.... C++ habit of not adding a comma to the end of the last array element, or you'll get an error.

**Comment by ryansmith94** [comment]
Why do you think they're duplicates @caperneoignis? The events are from two different modules so they're not duplicates 😄 

@jlowe64 I think this is ready to merge now, just waiting on integration tests for the new event in the plugin repo.

**Comment by caperneoignis** [comment]
@ryansmith94 Well because they are listed else where in the routing array. look at line 9 and line 19, then look at line 28 and line 30. They are the exact same. Sorry the code snippet I added are of two different keys that are listed twice in the routing array. But if they are for two different modules, how does the function differentiate between the two sets of identical keys? I mean under php, the keys just get overwritten if they are included more then once, and since the value and the key is the same, the overwritting does not cause any issues, at this time. But later on, if we change the value in the earlier instance of the key and not the later, the key will have the value overwritten. So that is where my confusion lies.

**Comment by ryansmith94** [comment]
Ahhhh I get you now, I was just confused by the snippet you sent, the duplicates can be removed.

**Comment by caperneoignis** [comment]
Okay, the keys and values were removed. Nothing else was changed.

**Comment by caperneoignis** [comment]
They put that close and comment way to close to the comment key I swear!

**Comment by ryansmith94** [comment]
I think this PR is ready to merge. I think @caperneoignis has also completed the integration tests in a PR for the plugin.

**Comment by jlowe64** [comment]
Thanks @ryansmith94 and @caperneoignis for your hard work!
